sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ServiceBusProcessorClient.cs

```diff
@@ -0,0 +1,982 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Diagnostics.CodeAnalysis;
+using System.Globalization;
+using System.Linq.Expressions;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core;
+using Azure.Messaging.ServiceBus.Core;
+using Azure.Messaging.ServiceBus.Diagnostics;
+using Microsoft.Azure.Amqp;
+
+namespace Azure.Messaging.ServiceBus
+{
+    /// <summary>
+    ///   A client responsible for reading <see cref="ServiceBusMessage" /> from a specific entity
+    ///   as a member of a specific consumer group.
+    ///
+    ///   A consumer may be exclusive, which asserts ownership over associated partitions for the consumer
+    ///   group to ensure that only one consumer from that group is reading the from the partition.
+    ///   These exclusive consumers are sometimes referred to as "Epoch Consumers."
+    ///
+    ///   A consumer may also be non-exclusive, allowing multiple consumers from the same consumer
+    ///   group to be actively reading events from a given partition.  These non-exclusive consumers are
+    ///   sometimes referred to as "Non-Epoch Consumers."
+    /// </summary>
+    ///
+    public class ServiceBusProcessorClient : IAsyncDisposable
+    {
+        private Func<ServiceBusMessage, ServiceBusSession, Task> _processMessage;
+        private Func<ExceptionReceivedEventArgs, Task> _processErrorAsync = default;
+        /// <summary>The primitive for synchronizing access during start and set handler operations.</summary>
+        private readonly object EventHandlerGuard = new object();
+        /// <summary>The primitive for synchronizing access during start and close operations.</summary>
+        private SemaphoreSlim MessageHandlerSemaphore;
+        /// <summary>The primitive for synchronizing access during start and close operations.</summary>
+        private readonly SemaphoreSlim ProcessingStartStopSemaphore = new SemaphoreSlim(1, 1);
+        private CancellationTokenSource RunningTaskTokenSource { get; set; }
+
+        private readonly ServiceBusReceiverClient _receiver;
```

Comment:
Should we consider the transport consumer here to get as close to the metal as possible, or is there functionality in the receiver that we'll need to build on?

Reply:
I think we will probably want another layer on top of the consumer and have both Processor/Receiver delegate to that. We can't just use the consumer because we need stuff from the Connection as well.